src/configure.ts

```diff
@@ -65,15 +73,60 @@ export async function jestPreviewConfigure(
     // }
   });
 
-  if (options.publicFolder) {
+  if (publicFolder) {
     createCacheFolderIfNeeded();
     fs.writeFileSync(
       path.join(CACHE_FOLDER, 'cache-public.config'),
-      options.publicFolder,
+      publicFolder,
       {
         encoding: 'utf-8',
         flag: 'w',
       },
     );
   }
 }
+
+function autoRunPreview() {
+  const originalIt = it;
+  const itWithPreview: jest.It = (name, callback, timeout) => {
+    let callbackWithPreview = undefined as jest.ProvidesCallback | undefined;
+    if (!callback) {
+      callbackWithPreview = undefined;
+    } else if (callback.constructor.name === 'AsyncFunction') {
+      console.log(callback.constructor.name);
+      callbackWithPreview = async function () {
+        try {
+          return await (callback as () => Promise<unknown>)();
```

Comment:
Opportunity to improve: We can write a helper function to detect an async function, that will make TypeScript happy and doesn't require type assertion here. I will add an example below.

Reply:
@thanhsonng I'm waiting for your example. You can commit directly as well if you want.

Reply:
Turns out it is not easy to achieve. We need to inspect the return value, i.e. we must call the `callback` function first.
On the other hand, checking `callback.constructor.name === 'AsyncFunction'` is not enough IMO. We actually want to check if the function return a Promise.

So I propose something like this. We always `await` for callback's return value, regardless of its signature, and we make `callbackWithPreview` always return a Promise.

```ts
if (!callback) {
  callbackWithPreview = undefined;
} else {
  callbackWithPreview = async (...args: Parameters<jest.ProvidesCallback>) => {
    try {
      // @ts-ignore
      return await callback(...args);
    } catch (error) {
      debug();
      throw error;
    }
  }
}
```

Reply:
> checking `callback.constructor.name === ‘AsyncFunction’` is not enough IMO. We actually want to check if the function return a Promise.

Can you elaborate why we need to check if the function return a Promise, instead of checking the function is an async function?


> So I propose something like this. We always `await` for callback’s return value, regardless of its signature, and we make `callbackWithPreview` always return a Promise.

I did try it before this and it works actually (not well tested yet). However, I have a few opening questions I haven’t had answers yet:
1. Is there any issues regarding the rightness when we always converting sync => async?
2. Is there any performance issues when we always use async (even when not needed)?

Reply:
>Can you elaborate why we need to check if the function return a Promise, instead of checking the function is an async function?

An async function is just syntactic sugar for functions that return a promise. Jest allows user to write functions that return a promise without using async/await syntax, so we must account for that case as well. This link is for reference: https://jestjs.io/docs/tutorial-async.

>I did try it before this and it works actually (not well tested yet). However, I have a few opening questions I haven’t had answers yet:
>1. Is there any issues regarding the rightness when we always converting sync => async?
>2. Is there any performance issues when we always use async (even when not needed)?

I believe those are non-issues but you shouldn't take my words for it. I will do some research about these questions. 😂

Reply:
Lemme do a benchmark.

Reply:
@thanhsonng I did a benchmark and the running time is very similar. So I will go with `always async` method.